File: lib/alien_demo_web/channels/user_socket.ex
defmodule AlienDemoWeb.UserSocket do
  use Phoenix.Socket

  channel "users:*", AlienDemoWeb.UserChannel
  channel "room:*", AlienDemoWeb.ChatRoomChannel
  channel "private:*", AlienDemoWeb.ChatPrivateChannel

  def connect(%{"user_token" => user_token}, socket, _connect_info) do
    with {:ok, user_id} <- login(user_token) do
      assign(socket, :user_id, user_id)

      {:ok, socket}
    else
      {:error, :unauthenticated} -> :error
      {:error, _} -> :error
    end
  end

  def id(socket), do: "user_socket:user:#{socket.assigns[:user_id]}"

  defp login(nil) do
    {:error, :unauthenticated}
  end

  defp login(user_token) do
    with {:ok, user_id} <-
           Phoenix.Token.verify(AlienDemoWeb.Endpoint, "e5fg", user_token, max_age: 86400) do
      {:ok, user_id}
    else
      error -> error
    end
  end
end

File: assets/js/user_socket.js
// NOTE: The contents of this file will only be executed if
// you uncomment its entry in "assets/js/app.js".

// To use Phoenix channels, the first step is to import Socket,
// and connect at the socket path in "lib/web/endpoint.ex".
//
// Pass the token on params as below. Or remove it
// from the params if you are not using authentication.
import { Socket } from "phoenix";

const socket = new Socket("/sockets/users", {
  params: { user_token: window.userToken }
});

socket.connect();

export default socket;

// Now that you are connected, you can join channels with a topic:
const userChannel = socket.channel("users:join");
userChannel
  .join()
  .receive("ok", resp => {
    console.log("Joined successfully", resp);
  })
  .receive("error", resp => {
    console.log("Unable to join", resp);
  });

// Now that you are connected, you can join channels with a topic:
const chatLobbyChannel = socket.channel("room:lobby", {
  user_id: window.userId
});
chatLobbyChannel
  .join()
  .receive("ok", resp => {
    console.log("Joined lobby successfully", resp);
  })
  .receive("error", resp => {
    console.log("Unable to join", resp);
  });

export { chatLobbyChannel, userChannel };
